mindsdb/integrations/libs/base.py

```diff
@@ -110,7 +110,75 @@ def __init__(self, name: str):
 
 class PredictiveHandler(BaseHandler):
     """
+    DEPRECATED.
+
     Base class for handlers associated to predictive systems.
     """
     def __init__(self, name: str):
         super().__init__(name)
+
+
+class BaseMLEngine:
+    """
+    Base class for integration engine to connect with other Machine Learning libraries/frameworks.
+
+    This class will be instanced when interacting with the underlying framework.
+    """
+
+    def __init__(self, model_storage, engine_storage) -> None:
+        """
+        Warning: This method should not be overridden.
+
+        Initialize storage objects required by the ML engine.
+
+        - engine_storage: persists global engine-related internals.
+        - model_storage: stores artifacts for any given model.
+        """
+        self.model_storage = model_storage
+        self.engine_storage = engine_storage
+
+    def create(self, target: str, df: Optional = Union[None, pd.DataFrame], args: Optional = dict) -> None:
+        """
+        Saves a model inside the engine registry for later usage.
+
+        Normally, an input dataframe is required to train the model.
+        However, some integrations may merely require registering the model instead of training, in which case `df` can be omitted.
+
+        Any other arguments required to register the model can be passed in an `args` dictionary.
+        """
+        raise NotImplementedError
+
+    def predict(self, df: pd.DataFrame, args: Optional = dict) -> pd.DataFrame:
+        """
+        Calls a model with some input dataframe `df`, and optionally some arguments `args` that may modify the model behavior.
+
+        The expected output is a dataframe with the predicted values in the target-named column.
+        Additional columns can be present, and will be considered row-wise explanations if their names finish with `_explain`.
+        """
+        raise NotImplementedError
+
+    def create_engine(self, connection_args: dict):
+        """
+        Optional.
+
+        Used to connect with external sources (e.g. a REST API) that the engine will require to use any other methods.
+        """
+        raise NotImplementedError
+
+    def update(self, df: Optional = Union[None, pd.DataFrame]) -> None:
```

Comment:
`Union[None, pd.DataFrame]` -> `Optional[pd.DataFrame]`

Reply:
Yeah, that's a fair concern, we can make it optional as in `create`.